# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [2]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [3]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [5]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [6]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [7]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [8]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [9]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [10]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [11]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [12]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [13]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [14]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [15]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [16]:
data.duplicated().sum()

71

In [17]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [18]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [19]:
data['total_income_category'] = data['total_income'].apply(categorize_income)
data.head(10)

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0     926.185831         27    высшее             0   
6         0    2879.202052         43    высшее             0   
7         0     152.779569         50   среднее             1   
8         2    6929.865299         35    высшее             0   
9         0    2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   
5  гражданский брак                 1      M   компаньон     0        255763   
6   женат / замужем                 0      F   компаньон     0        240525   
7   женат / замужем                 0      M   сотрудник     0        135823   
8  гражданский брак                 1      F   сотрудник     0         95856   
9   женат / замужем                 0      M   сотрудник     0        144425   

                      purpose total_income_category  
0               покупка жилья                     B  
1     приобретение автомобиля                     C  
2               покупка жилья                     C  
3  дополнительное образование                     B  
4             сыграть свадьбу                     C  
5               покупка жилья                     B  
6           операции с жильем                     B  
7                 образование                     C  
8       на проведение свадьбы                     C  
9     покупка жилья для семьи                     C

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [20]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [21]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [22]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [23]:
# Для нахождения зависимости между количеством детей и возвратом кредита в срок создадим сводную таблицу.

data_pivot_children = data.pivot_table(index=['children'], \
                                       columns='debt', \
                                       values='purpose', \
                                       aggfunc='count')

# Заполним пропуски нулем и приведем столбцы к типу integer.

data_pivot_children = data_pivot_children.fillna(0)
data_pivot_children[0] = data_pivot_children[0].astype('int')
data_pivot_children[1] = data_pivot_children[1].astype('int')

# Найдем долю должников от общего количества взявших кредит. Для этого создадим новый столбец 'debtors_rate'.

#data_pivot_children['debtors_rate'] = 100 * data_pivot_children[1] / (data_pivot_children[0] + \
#                                                                      data_pivot_children[1])

data_pivot_children['debtors_rate'] = 100 * data_pivot_children[1] / (data_pivot_children[0] + data_pivot_children[1])
# Выведем финальную сводную таблицу.

data_pivot_children

debt          0     1  debtors_rate
children                           
0         13028  1063      7.543822
1          4364   444      9.234609
2          1858   194      9.454191
3           303    27      8.181818
4            37     4      9.756098
5             9     0      0.000000

-

**Вывод**: По сводной таблице заметим, что самая низкая доля должников у клиентов без детей (7.5%). В семьях, где есть дети, количество должников в среднем повысилось на 1.5%, самая большая доля приходится на заемщиков с 4 детьми (9.7%). Выборку из 5 детей нельзя считать реперезентивной, так как слишком мало данных. Очевидно, что многодетным семьям сложнее выплачивать кредит в срок, скорее всего, из-за больших трат на содержание детей.

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [24]:
# Для нахождения зависимости между семейным положением и возвратом кредита в срок создадим сводную таблицу.

data_pivot_status = data.pivot_table(index=['family_status'], \
                                     columns='debt', \
                                     values='purpose', \
                                     aggfunc='count')

# Найдем долю должников от общего количества взявших кредит. Для этого создадим новый столбец 'debtors_rate'.

data_pivot_status['debtors_rate'] = 100 * data_pivot_status[1] / (data_pivot_status[0] + data_pivot_status[1])

# Выведем финальную сводную таблицу.

data_pivot_status

debt                       0    1  debtors_rate
family_status                                  
Не женат / не замужем   2523  273      9.763948
в разводе               1105   84      7.064760
вдовец / вдова           888   63      6.624606
гражданский брак        3749  385      9.313014
женат / замужем        11334  927      7.560558

-

**Вывод**: По сводной таблице заметим, что самая низкая доля должников у клиентов "вдовец/вдова" и "в разводе" (6.6% и 7.0% соответсвенно). Люди со статусом "не женат/не замужем" и "гражданский брак" являются самыми ненадежными замещиками (9.7% и 9.3% соответсвенно). Было бы интересно посмотреть группировку по возрасту, можем предположить, что клиенты "не женат/не замужем" и "гражданский брак" более молодые и менее ответсвенные.

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [25]:
# Переименуем категории в таблице data для наглядности.
# Пусть A - Сверхвысокий доход, B - Высокий доход, C - Доход выше среднего, D - Средний доход, E - Низкий доход.

def categorize_income_name(income):
    try:
        if income == 'A':
            return 'Сверхвысокий доход'
        elif income == 'B':
            return 'Высокий доход'
        elif income == 'C':
            return 'Доход выше среднего'
        elif income == 'D':
            return 'Средний доход'
        elif income == 'E':
            return 'Низкий доход'
    except:
        pass

data['total_income_category'] = data['total_income_category'].apply(categorize_income_name)

# Для нахождения зависимости между уровнем дохода и возвратом кредита в срок создадим сводную таблицу.

data_pivot_income = data.pivot_table(index=['total_income_category'], \
                                     columns='debt', \
                                     values='purpose', \
                                     aggfunc='count')

# Найдем долю должников от общего количества взявших кредит. Для этого создадим новый столбец 'debtors_rate'.

data_pivot_income['debtors_rate'] = 100 * data_pivot_income[1] / (data_pivot_income[0] + data_pivot_income[1])

# Выведем финальную сводную таблицу.

data_pivot_income

debt                       0     1  debtors_rate
total_income_category                           
Высокий доход           4660   354      7.060231
Доход выше среднего    14568  1353      8.498210
Низкий доход              20     2      9.090909
Сверхвысокий доход        23     2      8.000000
Средний доход            328    21      6.017192

**Вывод:** Данные о людях чей доход низок или сверхвысок не являются реперезентивными. Самая большая доля должников приходится на людей, чей доход выше среднего. Самая низкая доля - клиенты со средним доходом. Возможно люди с меньшими финансовыми возможностями более тщательно планируют свою бюджет.

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [26]:
# Для нахождения зависимости между целью кредита и возвратом кредита в срок создадим сводную таблицу.

data_pivot_purpose = data.pivot_table(index=['purpose_category'], \
                                      columns='debt', \
                                      values='purpose', \
                                      aggfunc='count')

# Найдем долю должников от общего количества взявших кредит. Для этого создадим новый столбец 'debtors_rate'.

data_pivot_purpose['debtors_rate'] = 100 * data_pivot_purpose[1] / (data_pivot_purpose[0] + data_pivot_purpose[1])

# Выведем финальную сводную таблицу.

data_pivot_purpose

debt                         0    1  debtors_rate
purpose_category                                 
операции с автомобилем    3879  400      9.347978
операции с недвижимостью  9971  780      7.255139
получение образования     3619  369      9.252758
проведение свадьбы        2130  183      7.911803

**Вывод:** Самые высокие доли должников приходятся на клиентов, у которых цель кредита - получение образования или покупка автомобиля. Получающие образование люди скорее всего являются студентами и финансово не могут оплатить кредит вовремя. Операции с недвижимостью показывают самую низкую долю, возможно люди берущие ипотеку имеют больший доход или более отвественные, ведь ипотека достаточно рисковый вид кредита. 

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* Пропуски были в столбцах 'days_employed' и 'total_income'. Возможно эти клиенты никогда не работали или работали неофициально.



#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Среднее значение может некорректно характеризовать данные из-за слишком выделяющихся значений. С помощью медианы мы объективно оценим информацию даже при наличии выбросов.

### Шаг 4: общий вывод.

-


Мы провели исследование, чтобы выявить влияние различных факторов на погашение кредита в срок и дать рекомендации банку. Входные данные — статистика о платёжеспособности клиентов.

Перед проведением исследования были поставлены следующие вопросы:

* Есть ли зависимость между количеством детей и возвратом кредита в срок?
* Есть ли зависимость между семейным положением и возвратом кредита в срок?
* Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
* Как разные цели кредита влияют на его возврат в срок?

**Вопрос 1:**

* 7.5% - доля должников среди заемщиков без детей. 
* 9.2% - доля должников среди заемщиков с 1 ребенком. 
* 9.4% - доля должников среди заемщиков с 2 детьми. 

Рекомендация: для должников с более чем 2 детьми следует провести дополнительные исследования, так как выборка слишком мала.

**Вопрос 2:**

* самая низкая доля должников у клиентов "вдовец/вдова" и "в разводе" (6.6% и 7.0% соответсвенно). 
* Люди со статусом "не женат/не замужем" и "гражданский брак" являются самыми ненадежными замещиками (9.7% и 9.3% соответсвенно)

Рекомендация: разница между первым и вторым пунктом составляет примерно 3%, что может критично отразиться на кредитном портфеле банка. 

**Вопрос 3:**

* 8.4% - доля должников среди заемщиков с доходом выше среднего. 
* 7.0% - доля должников среди заемщиков с высоким доходом. 

Рекомендация: для клиентов с другим уровнем дохода следует провести дополнительные исследования, т.к. данные не репрезентивны.

**Вопрос 4:**

* 9.3% - доля должников, целью которых были операции с автомобилем.
* 9.2% - доля должников, целью которых были операции с образованием.
* 7.9% - доля должников, целью которых были операции со свадьбой.
* 7.2% - доля должников, целью которых были операции с недвижимостью.

Рекомендация: разница между первыми и последними двумя пунктами составляет примерно 2%, что может критично отразиться на кредитном портфеле банка. 

#### Общие рекомендации:
1. Следует обратить внимание на зависимости из 2 и 4 вопроса (семейное положение и цель кредита)
2. Для первого и третьего вопроса следует провести дополнительные исследования
